dev/conductor/core/lib/src/git.dart

```diff
@@ -74,15 +74,55 @@ class Git {
     if ((result.stderr as String).isNotEmpty) {
       message.writeln('stderr from git:\n${result.stderr}\n');
     }
-    throw GitException(message.toString());
+    throw GitException(message.toString(), args);
   }
 }
 
+enum GitExceptionType {
+  /// Git push failed because the remote branch contained commits the local did
+  /// not.
+  ///
+  /// Either the local branch was wrong, and needs a rebase before pushing
+  /// again, or the remote branch needs to be overwritten with a force push.
+  ///
+  /// Example output:
+  ///
+  /// ```
+  /// To github.com:user/engine.git
+  ///
+  ///  ! [rejected]            HEAD -> cherrypicks-flutter-2.8-candidate.3 (non-fast-forward)
+  /// error: failed to push some refs to 'github.com:user/engine.git'
+  /// hint: Updates were rejected because the tip of your current branch is behind
+  /// hint: its remote counterpart. Integrate the remote changes (e.g.
+  /// hint: 'git pull ...') before pushing again.
+  /// hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+  /// ```
+  PushRejected,
+}
+
+/// An exception created because a git subprocess failed.
+///
+/// Known git failures will be assigned a [GitExceptionType] in the [type]
+/// field. If this field is null it means and unknown git failure.
 class GitException implements Exception {
-  GitException(this.message);
+  GitException(this.message, this.args) {
+    if (_pushRejectedPattern.hasMatch(message)) {
+      type = GitExceptionType.PushRejected;
+    } else {
+      // because type is late final, it must be explicitly set before it is
+      // accessed.
+      type = null;
+    }
+  }
+
+  static final RegExp _pushRejectedPattern = RegExp(
```

Comment:
Is there a way to detect this error by exit code instead of message? using something like exit code can make this code more deterministic.

Reply:
Unfortunately not. This particular error is exit 1, as are almost all errors, per: https://stackoverflow.com/questions/4917871/does-git-return-specific-return-error-codes